### ISSUE TITLE
Contradictory behavior when used $ref

### ISSUE THREAD

**Issue by habibutsu** [for-GET/jesse#13]
Working example:

``` erlang
ok = jesse:add_schema("./schema3", #{
    <<"id">> => <<"schema3">>,
    <<"type">> => <<"object">>,
    <<"properties">> => #{
        <<"key3">> => #{
            <<"type">> => <<"string">>
        }
    }
}).
ok = jesse:add_schema("./schema2", #{
    <<"id">> => <<"schema2">>,
    <<"type">> => <<"object">>,
    <<"properties">> => #{
        <<"key2">> => #{
            <<"allOf">> => [
                #{<<"$ref">> => <<"schema3">>}
            ]
        }
    }
}).
ok = jesse:add_schema("./schema1", #{
    <<"id">> => <<"schema1">>,
    <<"properties">> => #{
        <<"key1">> => #{
            <<"allOf">> => [
                #{<<"$ref">> => <<"schema2">>}
            ]
        }
    }
}).
{ok, _} =  jesse:validate("./schema1", #{<<"key1">> => #{<<"key2">> => #{<<"key3">> => <<"123">>}}}, [
    {default_schema_ver, <<"http://json-schema.org/draft-04/schema#">>}
]).
```

Problems:
1. Name of schema does not match to name that should be in **$ref**
2. If to delete **id** from schemas - example does not work
3. The items _1_ and _2_ is not valid and example works in simple manner (withot **id** and with clear naming) when used only two levels of nested.

**Comment by andreineculau** [comment]
Thanks for the report.

The issue is n-fold and it will take a while to get right AFAICS. Meanwhile, it would be great if you could submit a PR with new tests added to `json_validator_tests.erl` that you think should pass.

- [ ] normalize keys, ids, refs
- [ ] keys (e.g. `add_schema(Key,...`) should be FQDN URIs. Note: we could default to the file URI scheme, given a root folder
- [ ] id should default to Key, or the FQDN $ref

**Comment by seriyps** [comment]
I use another pattern, but meet a similar problems:

types.json
```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "definitions": {
        "myType1": {},
        "myType2": {}
    }
}
```

my_schema.json
```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "type": "object",
    "properties": {
        "my_field1": {"$ref": "types.json#definitions/myType1"},
        "my_field1": {"$ref": "types.json#definitions/myType1"}
    }
}
```

```
jesse:add_schema(<<"types.json">>, TypesAsJsonStruct),
jesse:add_schema(<<"my_schema.json">>, MySchemaAsJsonStruct),

jesse:validate(<<"my_schema.json">>, #{<<"my_field1">> => <<"my_value">>}),
```

And found that It performs lookup like `jesse_database:read("./types.json")`.

While the fact that it passes string, not binary, is more or less normal (but hard to debug), but it's also prepending `"./"` to name.
I've found (by tracing) that it's because it calls `"./" = filename:dirname("types.json")` inside `jesse_state:combine_relative_id("types.json","types.json#/definitions/myType1")`.

I don't know if this is some expected behaviour or not, but I spend a lot of time investigating this.

Resolved like that:

```
jsonschema_loader(Name) when is_list(Name) ->
    jsonschema_loader(list_to_binary(Name));
jsonschema_loader(<<"./", Name/binary>>) ->
    jsonschema_loader(Name);
jsonschema_loader(Name) ->
    jesse_database:read(Name).

my_fun() ->
    Opts = [{schema_loader_fun, fun jsonschema_loader/1}],
    jesse:validate(<<"my_schema.json">>, #{<<"my_field1">> => <<"my_value">>}, Opts).
```

**Comment by marutha** [comment]
+1

**Comment by andreineculau** [comment]
@marutha try running the jesse version in PR #24 and report back if it helps. thanks!

**Comment by andreineculau** [comment]
fixed by #24
